src/Output/Copy/Serializer/ContentType/XmlContentTypeSerializer.php

```diff
@@ -0,0 +1,373 @@
+
+        switch ($node->nodeType) {
+            case XML_CDATA_SECTION_NODE:
+                $output[$this->config['cdataKey']] = $this->normalizeTextContent($node->textContent);
+                break;
+
+            case XML_TEXT_NODE:
+                $output = $this->normalizeTextContent($node->textContent);
+                break;
+
+            case XML_ELEMENT_NODE:
+                $output = $this->parseChildNodes($node, $output);
+                $output = $this->normalizeNodeValues($output);
+                $output = $this->collectAttributes($node, $output);
+                break;
+        }
+
+        return $output;
+    }
+
+    protected function parseChildNodes(DOMNode $node, $output)
+    {
+        foreach ($node->childNodes as $child) {
+            if ($child->nodeType === XML_CDATA_SECTION_NODE) {
+                if (!is_array($output)) {
+                    if (!empty($output)) {
+                        $output = [$this->config['valueKey'] => $output];
+                    } else {
+                        $output = [];
+                    }
+                }
+
+                $output[$this->config['cdataKey']] = $this->normalizeTextContent($child->textContent);
+            } else {
+                $value = $this->parseNode($child);
+
+                if ($child->nodeType == XML_TEXT_NODE) {
+                    if ($value != '') {
+                        if (!empty($output)) {
+                            $output[$this->config['valueKey']] = $value;
+                        } else {
+                            $output = $value;
+                        }
+                    }
+                } elseif ($child->nodeType !== XML_COMMENT_NODE) {
+                    $nodeName = $child->nodeName;
+
+                    if (!isset($output[$nodeName])) {
+                        $output[$nodeName] = [];
+                    }
+
+                    $output[$nodeName][] = $value;
+                }
+            }
+        }
+
+        return $output;
+    }
+
+    protected function normalizeTextContent($textContent): string
+    {
+        return trim(
+            preg_replace(
+                [
+                    '/\n+\s+/',
+                    '/\r+\s+/',
+                    '/\n+\t+/',
+                    '/\r+\t+/'
+                ],
+                ' ',
+                $textContent
+            )
+        );
+    }
+
+    protected function normalizeNodeValues($values)
+    {
+        if (is_array($values)) {
+            // if only one node of its kind, assign it directly instead if array($value);
+            foreach ($values as $key => $value) {
+                if (is_array($value) && count($value) === 1) {
+                    $keyName = array_keys($value)[0];
+
+                    if (is_numeric($keyName)) {
+                        $values[$key] = $value[$keyName];
+                    }
+                }
+            }
+
+            if (empty($values)) {
```

Comment:
you can check this before the foreach because you're not removing elements from $values or smth like that... you could also do some early return like
```php
if (!is_array($values)) {
    return $values;
}
if (empty($values)) {
    return '';
}

foreach (...) {
    ...
}
return $values;
```